### ISSUE TITLE
Build monero-java and monero-cpp to static libraries with shared library wrapper [2 XMR]

### ISSUE THREAD

**Issue by woodser** [haveno-dex/haveno#795]
This issue requests building [monero-java](https://github.com/woodser/monero-java) and [monero-cpp](https://github.com/woodser/monero-cpp) to static libraries with a shared library wrapper.

For example: libmonero-cpp.a, libmonero-java.a, and libmonero-java.so.

This will allow us to switch from using monero-wallet-rpc to native wallet bindings in the future for better performance and scalability.

This issue shares the same bounty as the related issue in monero-java: https://github.com/woodser/monero-java/issues/74

**Comment by woodser** [comment]
The dynamic libraries currently built will work with JDK > 11.

**Comment by niyid** [comment]
I just saw this had been reopened. What is the issue?

**Comment by woodser** [comment]
I think the only issue is that the dynamic libraries require MSYS2 to be installed and on the Path variable in Windows, whereas static libraries would be self-contained and shouldn't require external dependencies.

**Comment by woodser** [comment]
No, creating static libraries with a shared library wrapper is the request of this issue.

**Comment by woodser** [comment]
The libraries should not have external dependencies which need to be installed, but should be standalone, which switching to static libraries ought to support?

**Comment by mcneb10** [comment]
I would like to look into this

**Comment by woodser** [comment]
@mcneb10 Great!

**Comment by woodser** [comment]
@mcneb10 Any progress?

Currently, the dynamic libraries built for Linux only work on Ubuntu 20.0.4 as far as I know (what they're built from). But there's a linker error on later versions.

This would be resolved by building to static libraries.

**Comment by woodser** [comment]
Bounty bumped to 3 XMR.

**Comment by mcneb10** [comment]
I'm sorry I have been distracted and unmotivated recently. I was not able to compile to source due to some linking errors with the unbound package. I will have another look today.

**Comment by woodser** [comment]
No worries, we can leave this unassigned if it doesn't work out.

**Comment by woodser** [comment]
@mcneb10 Please ping me if I can help with the linker error. I've built it a few times :)

**Comment by woodser** [comment]
Not quite. That issue is to create universal binaries which run natively on mac x86_64 and arm64 without emulation, whereas this issue is to build the monero-java libs to static libraries with a dynamic wrapper for all platforms, for portability.

**Comment by mcneb10** [comment]
Note to anyone compiling this: if you try to compile on anything except a distro based on stable debian you will run into errors very quickly

**Comment by nsec1** [comment]
@woodser , I need help to integrate new monero-java.so to haveno. Do I just have to "mvn install" on monero-java and then build haveno with make?

**Comment by woodser** [comment]
I believe you can simply add the path to the newly built libraries to your `CLASSPATH` variable and then start Haveno with `--useNativeXmrWallet=true`.

But to do a complete install:

- Bump 0.8.31 to 0.8.32 in monero-java's pom.xml
- In monero-java: `mvn clean deploy -P release -Dmaven.deploy.skip=true`
- Bump `moneroJavaVersion` to `0.8.32` in Haveno's build.gradle
- Uncomment `mavenLocal()` in build.gradle
- Rebuild Haveno, updating the verification metadata for the new dependency: `./gradlew --write-verification-metadata sha256 && ./gradlew build --refresh-keys --refresh-dependencies -x test -x checkstyleMain -x checkstyleTest`
- Start Haveno with `--useNativeXmrWallet=true`

You could make some change in monero-java to confirm the update however you do it, e.g. adding `std::cout << "getting balance!" << std::endl;` to `uint64_t monero_wallet_full::get_balance() const` in monero_wallet_full.cpp.

**Comment by nsec1** [comment]
@woodser , could you check?
woodser/monero-cpp#64
woodser/monero-java#105

**Comment by nsec1** [comment]
@woodser, could you check woodser/monero-java#105?

**Comment by woodser** [comment]
> Is this still outstanding? Anyone currently working on it?

It's being worked on by @nsec1 here: https://github.com/woodser/monero-java/pull/105

**Comment by KewbitXMR** [comment]
Is this actively being worked on?

**Comment by woodser** [comment]
Yes, @nsec1 has new change I need to test: https://github.com/woodser/monero-java/pull/105#issuecomment-2303096985

**Comment by U65535F** [comment]
It's been a while and what's the progress till now?

**Comment by U65535F** [comment]
@woodser ping. Still being worked on?

**Comment by U65535F** [comment]
@woodser ping 2. Still being worked on?

**Comment by woodser** [comment]
Yes, I still need to test the latest solution, and it will still require more work. It just hasn't been a priority lately.

**Comment by NorrinRadd** [comment]
https://github.com/woodser/monero-cpp/pull/80 and https://github.com/woodser/monero-java/pull/120 address this issue.

**Comment by chrisplusplus** [comment]
Is this still outstanding ?  I would like to tackle it

**Comment by NorrinRadd** [comment]
It’s finished.

**Comment by fxrstor** [comment]
@woodser Close this issue as well.
